service/dataset.py

```diff
@@ -1,34 +1,109 @@
+                            "schema": '{"fake":"schema"}'
+                        },
+                        "budget":3.0}}
 
-DATASETS = {"example": {"type": "local_csv",
-                        "local_path": os.path.join(os.path.dirname(__file__), "datasets", "example.csv"),
-                        "key": "csv_details"},
-            "demo_dataverse": {"type": "dataverse",
-                               "local_metadata_path": os.path.join(os.path.dirname(__file__),
-                                                                   "datasets",
-                                                                   "dataverse",
-                                                                   "demo_dataverse.yml"),
-                               "key": "dataverse_details",
-                               "host": "https://demo.dataverse.org/api/access/datafile/395811"}}
 
+KNOWN_DATASET_TYPE_KEYS = ["csv_details", "dataverse_details"]
 
-def read(info):
+def read(dataset_request):
     """Get information needed to load the dataset
 
     :param info: The dataset to read and budget to use.
     :type info: dict {"dataset_name": str, "budget":int}
     :return: A dataset document that contains the type and info of the dataset
     :rtype: dict{"dataset_type": str, dataset_key: dict}
     """
-    dataset_name = info["dataset_name"]
+    dataset_name = dataset_request["dataset_name"]
 
     if dataset_name not in DATASETS:
         abort(400, "Dataset id {} not found.".format(dataset_name))
+    
     dataset = DATASETS[dataset_name]
-    if dataset["type"] == "dataverse":
-        dataset["token"] = secrets_get(name="dataverse:{}".format(info["dataset_name"]))["value"]
 
-    return {"dataset_type": dataset["type"], dataset["key"]: dataset}
+    # Validate the secret, extract token
+    if dataset["dataset_type"] == "dataverse_details":
+        dataset["dataset_type"]["token"] = secrets_get(name="dataverse:{}".format(dataset_request["dataset_name"]))["value"]
+
+    # Check/Decrement the budget before returning dataset
+    adjusted_budget = dataset["budget"] - dataset_request["budget"]
+    if adjusted_budget >= 0.0:
+        dataset["budget"] = adjusted_budget
+    else:
+        abort(412, "Not enough budget for read. Remaining budget: {}".format(dataset_name))
+
+    return dataset
+
+#%%
+def register(dataset):
+    dataset_name = dataset["dataset_name"]
+
+    if dataset_name in DATASETS:
+        abort(401, "Dataset id {} already exists. Identifies must be unique".format(dataset_name))
+
+    # Add key if possible
+    if dataset["dataset_type"] not in KNOWN_DATASET_TYPE_KEYS:
+        abort(402, "Given type was {}, must be either csv_details or dataverse_details.".format(str(dataset["dataset_type"])))
+    
+    # Add budget if possible 
+    if dataset["budget"]:
+        b = dataset["budget"]
+        if b <= 0.0: abort(403, "Budget must be greater than 0.")
```

Comment:
we don't usually use multi line statements. Not sure why this b = dict lookup. dict lookup = b thing is happening

Reply:
lol whoops